Sources/MongoSwift/ChangeStream.swift

```diff
@@ -0,0 +1,174 @@
+public struct ChangeStreamToken: Codable {
+    private let resumeToken: Document
+
+    public init(resumeToken: Document) {
+        self.resumeToken = resumeToken
+    }
+
+    public init(from decoder: Decoder) throws {
+      self.resumeToken = try Document(from: decoder)
+    }
+}
+
+/// A MongoDB ChangeStream.
+/// - SeeAlso: https://docs.mongodb.com/manual/changeStreams/
+public class ChangeStream<T: Codable>: Sequence, IteratorProtocol {
+  /// The cached resume token.
+  public private(set) var resumeToken: ChangeStreamToken
+
+  /// A `MongoClient` stored to make sure the source client stays valid
+  /// until the change stream is destroyed.
+  private var client: MongoClient?
+
+  /// A `ClientSession` stored to make sure the session stays valid
+  /// until the change stream is destroyed.
+  private var session: ClientSession?
+
+  /// A reference to the `mongoc_change_stream_t` pointer.
+  private var changeStream: OpaquePointer?
+
+  /// Used for storing Swift errors.
+  private var swiftError: Error?
+
+  /// The error that occurred while iterating the ChangeStream cursor,
+  /// if one exists. This should be used to check for errors after
+  /// `next()` returns `nil`.
+  public var error: Error? {
+    if let err = self.swiftError {
+        return err
+    }
+
+    var replyPtr = UnsafeMutablePointer<BSONPointer?>.allocate(capacity: 1)
+    defer {
+        replyPtr.deinitialize(count: 1)
+        replyPtr.deallocate()
+    }
+
+    var error = bson_error_t()
+    guard mongoc_cursor_error_document(self.changeStream, &error, replyPtr) else {
+        return nil
+    }
+
+    // If a reply is present, it implies the error occurred on the server. This *should* always be a commandError,
+    // but we will still parse the mongoc error to cover all cases.
+    if let docPtr = replyPtr.pointee {
+        // we have to copy because libmongoc owns the pointer.
+        let reply = Document(copying: docPtr)
+        return extractMongoError(error: error, reply: reply)
+    }
+
+    // Otherwise, the only feasible error is that the user tried to advance a dead change stream cursor,
+    // which is a logic error. We will still parse the mongoc error to cover all cases.
+    return extractMongoError(error: error)
+  }
+
+  /// Decoder for decoding documents into type `T`.
+  internal let decoder: BSONDecoder
+
+  /// Returns the next `T` in the change stream or nil if there is no next value.
+  /// Will block for a maximum of `maxAwaitTimeMS` milliseconds as specified in the
+  /// `ChangeStreamOptions`, or for the server default timeout if omitted.
+  public func next() -> T? {
+    let cursor = self.changeStream
+    // Allocate space for a reference to a BSON pointer.
+    let out = UnsafeMutablePointer<BSONPointer?>.allocate(capacity: 1)
+
+    // Check that there’s another document in the stream.
+    guard mongoc_change_stream_next(cursor, out) else {
+      return nil
+    }
+
+    defer {
```

Comment:
Let's move this defer block to immediately after the allocation. Otherwise, if the above guard fails we'll leak memory.